### PR TITLE
add validation middleware

### PR DESCRIPTION
**Pull request by vasylkushnir**

### CODE REVIEW

File: src/dto/task.dto.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+export class ReplaceTaskDto extends CreateTaskDto {}
+
+export class IdParam {
+  @IsMongoId({ message: 'invalid Id' })
+  id!: string
+}
+
+export class GetDeleteTaskDto extends IdParam {}
```
Comment (Ilchuk-Mihail) [inline src/dto/task.dto.ts:35]: `DeleteTaskDto` is better

File: src/middlewares/validationResults.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { ClassConstructor, plainToClass } from 'class-transformer'
+import { validate } from 'class-validator'
+import logger from '../lib/logger'
+import { ValidationError } from '../errors/HttpErrors'
+import { CreateTaskDto, ReplaceTaskDto, UpdateTaskDto, IdParam, GetDeleteTaskDto } from '../dto/task.dto'
+
+type SourceData = 'body' | 'query' | 'params'
+type TaskDto = CreateTaskDto | UpdateTaskDto | ReplaceTaskDto | ReplaceTaskDto | IdParam | GetDeleteTaskDto
+
+export default function validateRequest (input: ClassConstructor<TaskDto>, source: SourceData) {
```
Comment (Ilchuk-Mihail) [inline src/middlewares/validationResults.ts:11]: `ClassConstructor<any>` it's fine in this case have `any` type

File: src/middlewares/validationResults.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  return (req: Request, res: Response, next: NextFunction): void => {
+    const data = req[source]
+    const dtoObj = plainToClass(input, data)
+    validate(dtoObj).then(errors => {
+      if (errors.length > 0) {
+        const constraints = errors.map(obj => {
```
Comment (Ilchuk-Mihail) [inline src/middlewares/validationResults.ts:17]: I would rather merge all constraints arrays into one - > `const constraints = errors.map(({ constraints }: any) => Object.values(constraints)).flat()`

File: src/middlewares/validationResults.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+          return { constrains: obj.constraints }
+        })
+        next(new ValidationError({ constraints }))
+      } else {
+        logger.info('validation succeed')
```
Comment (Ilchuk-Mihail) [inline src/middlewares/validationResults.ts:22]: There is no need to log this.

File: src/routes.ts [hunk 5]
```diff
@@ -1,19 +1,20 @@
...
+  .post(validateRequest(CreateTaskDto, 'body'),
+    task.createTask)
   .get(task.getAllTasks)
 
 router.route('/tasks/:id')
+  .all(validateRequest(IdParam, 'params'))
   .get(task.getTaskById)
-  .put(task.replaceTask)
-  .patch(task.updateTask)
+  .put(validateRequest(UpdateTaskDto, 'body'), task.replaceTask)
```
Comment (Ilchuk-Mihail) [inline src/routes.ts:16]: ReplaceTaskDto ?

File: src/middlewares/errorHandler.ts [hunk 6]
```diff
@@ -5,12 +5,15 @@ import logger from '../lib/logger'
 export default function errorHandler (error: BaseError, req: Request, res: Response, next: NextFunction): void {
   const status = error.status || 500
   const message = error.message || 'Something went wrong '
+  const meta = error.meta || {}
 
   logger.error('Error Message:', {
```
Comment (Ilchuk-Mihail) [inline src/middlewares/errorHandler.ts:10]: You need to reuse error message as a logged message, also log meta directly and save a stack trace for debug purposes.

```
  logger.error(error.message, {
    ...meta,
    stack: error.stack
  })
```

File: src/middlewares/errorHandler.ts [hunk 7]
```diff
@@ -5,12 +5,15 @@ import logger from '../lib/logger'
...
     message: message,
-    statusCode: status
+    statusCode: status,
+    meta: meta
   })
   res.send({
```
Comment (Ilchuk-Mihail) [inline src/middlewares/errorHandler.ts:15]: Here you always sending  `200` :)

**Review by Ilchuk-Mihail — CHANGES_REQUESTED** [review]

File: src/dto/task.dto.ts (hunk 1 above)
Comment (vasylkushnir) [reply src/dto/task.dto.ts]: fixed
d05ad7787eac1bd92d123586f034466887fdb451

File: src/middlewares/validationResults.ts (hunk 2 above)
Comment (vasylkushnir) [reply src/middlewares/validationResults.ts]: fixed d05ad7787eac1bd92d123586f034466887fdb451

File: src/middlewares/validationResults.ts (hunk 3 above)
Comment (vasylkushnir) [reply src/middlewares/validationResults.ts]: fixed d05ad7787eac1bd92d123586f034466887fdb451

File: src/middlewares/validationResults.ts (hunk 4 above)
Comment (vasylkushnir) [reply src/middlewares/validationResults.ts]: fixed d05ad7787eac1bd92d123586f034466887fdb451

File: src/routes.ts (hunk 5 above)
Comment (vasylkushnir) [reply src/routes.ts]: fixed d05ad7787eac1bd92d123586f034466887fdb451

File: src/middlewares/errorHandler.ts (hunk 6 above)
Comment (vasylkushnir) [reply src/middlewares/errorHandler.ts]: fixed d05ad7787eac1bd92d123586f034466887fdb451

File: src/middlewares/errorHandler.ts (hunk 7 above)
Comment (vasylkushnir) [reply src/middlewares/errorHandler.ts:15]: fixed d05ad7787eac1bd92d123586f034466887fdb451

File: src/middlewares/errorHandler.ts [hunk 8]
```diff
@@ -5,12 +5,15 @@ import logger from '../lib/logger'
...
+  const meta = error.meta || {}
 
-  logger.error('Error Message:', {
-    message: message,
-    statusCode: status
+  logger.error(error.message,  {
+    ...meta,
+    stack: error.stack
   })
+  res.statusCode = status
```
Comment (Ilchuk-Mihail) [inline src/middlewares/errorHandler.ts:14]: Better to use same syntax like in controllers . `req.status(status).send(data)`

Comment (vasylkushnir) [reply src/middlewares/errorHandler.ts]: fixed 36b7a9d8d0f63934c3efff975d940c17659ee1c9

**Review by Ilchuk-Mihail — APPROVED** [review]

**Pull request merged**